File: platforms/web/core/engine.js
var Gamefic = (function() {
	var onStart = null;
	var onUpdate = null;

	return {
		start: function() {
			Opal.gvars.engine.$run();	
			var state = Opal.gvars.engine.$user().$character().$state();
			var json = state.$to_json();
			var response = JSON.parse(json);
			if (onStart) {
				onStart(response);
			}
		},

		receive: function(input) {
			Opal.gvars.engine.$receive(input);
			this.update();
		},

		update: function() {
			Opal.gvars.engine.$update();
			var state = Opal.gvars.engine.$user().$character().$state();
			var json = state.$to_json();
			var response = JSON.parse(json);
			if (onUpdate) {
				onUpdate(response);
			}
			if (Opal.gvars.engine.$user().$character().$queue().$length() > 0) {
				this.update();
			}
		},

		onStart: function(callback) {
			onStart = callback;
		},	

		onUpdate: function(callback) {
			onUpdate = callback;
		},
	}
})();

File: lib/gamefic-opal/engine.rb
# HACK: Explicit requires for Opal
require 'gamefic/world/playbook'
require 'gamefic/plot/darkroom'
require 'gamefic/grammar'
require 'gamefic/query'
require 'gamefic/grammar/verb_set'
require 'gamefic/grammar/conjugator'

module Gamefic
  module Opal
    class Engine < Gamefic::Engine
      attr_reader :user

      def post_initialize
        self.user_class = Gamefic::Opal::User
      end

      def run
        connect
        plot.introduce @user.character
        plot.ready
        @user.update
      end

      def turn
        @plot.ready
      end

      def receive input
        @user.character.queue.push input unless input.nil?
      end

      private

      def update
        @plot.update
        turn
      end
    end
  end
end
